docs/src/main/asciidoc/security.adoc

```diff
@@ -96,12 +98,32 @@ Additionally it provides `JWT Generation API` for creating `signed`, `inner-sign
 
 See the link:security-jwt[Using SmallRye JWT] guide for more information.
 
-=== OAuth2
+==== OAuth2
 
 `quarkus-elytron-security-oauth2` provides an alternative to `quarkus-oidc` Bearer Token Authentication Mechanism. It is based on `Elytron` and is primarily meant for introspecting the opaque tokens remotely.
 
 See the link:security-oauth2[Using OAuth2] guide for more information.
 
+==== Choosing between OpenId Connect, SmallRye JWT and OAuth2 extensions
+
+`quarkus-oidc` extension requires an OpenId Connect provider such as Keycloak which can be used to verify the Bearer tokens or authenticate the human users with the Authorization Code flow. In both cases `quarkus-oidc` requires a connection to this OpenId Connect provider.
+
+`quarkus-oidc` is the only option when the user authentication via Authorization Code flow or supporting multiple tenants is required.
+
+When the Bearer tokens have to be verified then `quarkus-oidc`, `quarkus-smallrye-jwt` and `quarkus-elytron-security-oauth2` can be used.
+
+If you have Bearer tokens in a JWT format then all these 3 extensions can be used. Both `quarkus-oidc` and `quarkus-smallrye-jwt` support refreshing the JWK set when the OpenId Connect provider rotates the keys, therefore `quarkus-oidc` or `quarkus-smallrye-jwt` should be used for verifying JWT tokens if the remote token introspection should be avoided or not supported by the providers.
+
+`quarkus-smallrye-jwt` does not support the remote introspection of the opaque tokens or even JWT tokens - it always relies on the locally available keys - possibly fetched from the OpenId Connect provider. So if you need to introspect the JWT tokens remotely then both `quarkus-oidc` and `quarkus-elytron-security-oauth2` will work. Both extensions also support the verification of the opaque/binary tokens via the remote introspection.
+
+`quarkus-oidc` and `quarkus-smallrye-jwt` can have both JWT and opaque tokens injected into the endpoint code - the injected JWT tokens may offer a richer information about the user.
+
+`quarkus-smallrye-jwt` supports more key formats than `quarkus-oidc`. The latter will only use the JWK-formatted keys which are part of a JWK set. The former - can also work with PEM keys and single JWK keys.
+
+`quarkus-smallrye-jwt` can handle locally not only signed but also inner-signed-and-encrypted or only encrypted tokens. In fact `quarkus-oidc` and `quarkus-elytron-security-oauth2` can verify such tokens too but only by treating them as opaque tokens and verifying them via the remote introspection.
+
+`quarkus-elytron-security-oauth2` is the best choice if you need a light weight library for the remote introspection of either JWT or opaque tokens.
```

Comment:
Grouping into sections also help. For example:

## quarkus-oidc

- Does not support X
- Does support Y
- Use it when Z

## quarkus-smallrye-jwt

- Does not support X
- Does support Y
- Use it when Z

## quarkus-elytron-security-oauth2

- Does not support X
- Does support Y
- Use it when Z

Reply:
@gastaldi thanks for the ideas, I've been thinking during a walk that I'd not mind to start with the prose - this is just an initial try, as I said, I'd like to have some deeper recommendations which would be tricky to catch with just some bullets/table items. There are already dedicated sibling subsections for each of these extensions just above this newly introduced section which some basic info about the differences, so we'd introduce a lower-level mirror with the subsections I guess. But I was liking your idea of having a table too, so I'd like to have this test + cheat sheet table capturing the main points from the text... Thanks very much for trying to help :-), I'll give a few options a try a bit later on :-)